src/modules/flow/netctl/netctl.c

```diff
@@ -0,0 +1,342 @@
+        SOL_FLOW_NODE_TYPE_NETCTL_LIST_SERVICES__OUT__OUT,
+        name);
+    if (r < 0)
+        SOL_WRN("Failed to send network service name: %s", name);
+
+    sol_ptr_vector_append(&mdata->service_list, service);
+}
+
+void
+error_cb(void *data, const struct sol_netctl_service *service,
+    unsigned int error)
+{
+    const char *name;
+
+    name = sol_netctl_service_get_name(service);
+
+    if (name)
+        SOL_ERR("Service %s error is %d", name, error);
+}
+
+static int
+open_list_services(struct sol_flow_node *node, void *data, const struct sol_flow_node_options *options)
+{
+    static const char *errmsg = "Could not add service monitor";
+    struct list_service_data *mdata = data;
+
+    mdata->node = node;
+
+    sol_ptr_vector_init(&mdata->service_list);
+
+    if (sol_netctl_add_service_monitor(network_service_cb, mdata))
+        goto error;
+
+    return 0;
+
+error:
+    sol_flow_send_error_packet(node, EINVAL, "%s", errmsg);
+    SOL_WRN("%s", errmsg);
+
+    return -EINVAL;
+}
+
+static void
+close_list_services(struct sol_flow_node *node, void *data)
+{
+    static const char *errmsg = "Could not delete service monitor";
+    struct list_service_data *mdata = data;
+
+    if (sol_netctl_del_service_monitor(network_service_cb, mdata))
+        goto error;
+
+    sol_ptr_vector_clear(&mdata->service_list);
+
+    return;
+
+error:
+    sol_flow_send_error_packet(node, EINVAL, "%s", errmsg);
+    SOL_WRN("%s", errmsg);
+}
+
+static int
+connection_handle(struct sol_flow_node *node, void *data, bool connect)
+{
+    struct connection_service_data *mdata = data;
+    const struct sol_ptr_vector *service_list;
+    struct sol_netctl_service *service;
+    uint16_t i = 0;
+    enum sol_netctl_service_state state;
+    int r;
+    const char *name;
+
+    static const char *state_msgs[] = {
+        [SOL_NETCTL_SERVICE_STATE_UNKNOWN] = "Unknown",
+        [SOL_NETCTL_SERVICE_STATE_IDLE] = "Idle",
+        [SOL_NETCTL_SERVICE_STATE_ASSOCIATION] = "Association",
+        [SOL_NETCTL_SERVICE_STATE_CONFIGURATION] = "Configuration",
+        [SOL_NETCTL_SERVICE_STATE_READY] = "Ready",
+        [SOL_NETCTL_SERVICE_STATE_ONLINE] = "Online",
+        [SOL_NETCTL_SERVICE_STATE_DISCONNECT] = "Disconnect",
+        [SOL_NETCTL_SERVICE_STATE_FAILURE] = "Failure",
+        [SOL_NETCTL_SERVICE_STATE_REMOVE] = "Remove",
+    };
+
+    service_list = sol_netctl_get_services();
+
+    while (service = sol_ptr_vector_get(service_list,i)) {
+        state = sol_netctl_service_get_state(service);
+        name = sol_netctl_service_get_name(service);
+
+        if (name && strcmp(name, mdata->service_name) == 0) {
+            if (state == SOL_NETCTL_SERVICE_STATE_IDLE && connect) {
+                sol_netctl_service_connect(service);
+            } else if ((state == SOL_NETCTL_SERVICE_STATE_READY) && (!connect)) {
+                  sol_netctl_service_disconnect(service);
+            }
+            state = sol_netctl_service_get_state(service);
```

Comment:
it is not correct to query the state here. The call is asynchronous, so it's likely to be the old state since DBus signal did not arrive yet.

The proper way is to remove this `get_state()` from here and listen for service change callback. Then you remember the last-emitted value (ie: the enum value) and if it changed, then you emit the new value and remember the new one.

That is independent from connect/disconnect requests.